api/src/test/java/org/pmiops/workbench/billing/BillingProjectBufferServiceTest.java

```diff
@@ -284,4 +323,75 @@ public void syncBillingProjectStatus_iteratePastStallingRequests() {
+    entry.setStatusEnum(AVAILABLE);
+    entry.setFireCloudProjectName("test-project-name");
+    entry.setCreationTime(new Timestamp(NOW.toEpochMilli()));
+    billingProjectBufferEntryDao.save(entry);
+
+    User user = mock(User.class);
+    doReturn("fake-email@aou.org").when(user).getEmail();
+
+    BillingProjectBufferEntry assignedEntry = billingProjectBufferService.assignBillingProject(user);
+
+    ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);
+    ArgumentCaptor<String> secondCaptor = ArgumentCaptor.forClass(String.class);
+    verify(fireCloudService).addUserToBillingProject(captor.capture(), secondCaptor.capture());
+    String invokedEmail = captor.getValue();
+    String invokedProjectName = secondCaptor.getValue();
+
+    assertThat(invokedEmail).isEqualTo("fake-email@aou.org");
+    assertThat(invokedProjectName).isEqualTo("test-project-name");
+
+    assertThat(billingProjectBufferEntryDao.findOne(assignedEntry.getId()).getStatusEnum()).isEqualTo(ASSIGNED);
+  }
+
+  @Test
+  @Transactional(propagation = Propagation.NOT_SUPPORTED)
+  public void assignBillingProject_locking() throws InterruptedException, ExecutionException {
+    BillingProjectBufferEntry firstEntry = new BillingProjectBufferEntry();
+    firstEntry.setStatusEnum(AVAILABLE);
+    firstEntry.setFireCloudProjectName("test-project-name-1");
+    firstEntry.setCreationTime(new Timestamp(NOW.toEpochMilli()));
+    billingProjectBufferEntryDao.save(firstEntry);
+
+    User firstUser = new User();
+    firstUser.setEmail("fake-email-1@aou.org");
+    userDao.save(firstUser);
+
+    BillingProjectBufferEntry secondEntry = new BillingProjectBufferEntry();
+    secondEntry.setStatusEnum(AVAILABLE);
+    secondEntry.setFireCloudProjectName("test-project-name-2");
+    secondEntry.setCreationTime(new Timestamp(NOW.toEpochMilli()));
+    billingProjectBufferEntryDao.save(secondEntry);
+
+    User secondUser = new User();
+    secondUser.setEmail("fake-email-2@aou.org");
+    userDao.save(secondUser);
+
+    doAnswer(new CallsRealMethodsWithDelay(500)).when(billingProjectBufferEntryDao).save(any(BillingProjectBufferEntry.class));
+
+    Callable<BillingProjectBufferEntry> t1 = () -> billingProjectBufferService.assignBillingProject(firstUser);
+    Callable<BillingProjectBufferEntry> t2 = () -> billingProjectBufferService.assignBillingProject(secondUser);
+
+    List<Callable<BillingProjectBufferEntry>> callableTasks = new ArrayList<>();
+    callableTasks.add(t1);
+    callableTasks.add(t2);
+
+    ExecutorService executor = Executors.newFixedThreadPool(2);
+    List<Future<BillingProjectBufferEntry>> futures = executor.invokeAll(callableTasks);
+
+    assertThat(futures.get(0).get().getFireCloudProjectName())
+        .isNotEqualTo(futures.get(1).get().getFireCloudProjectName());
+
+    // Need to clean up entities since they actually get committed to the database because we're not running within a transaction
```

Comment:
This is the test annotation that handles cleanup:

```
@DirtiesContext(classMode = ClassMode.BEFORE_EACH_TEST_METHOD)
```

Also I don't see Propagation.NOT_SUPPORTED on this test - is it really not being cleaned up or was this comment from an earlier version?

Reply:
I added it as an annotation.

The dirties context annotation also works! Thanks for the tip